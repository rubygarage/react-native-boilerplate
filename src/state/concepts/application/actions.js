import * as types from './types';

export const setOnboardingCompletion = (isCompleted) => ({
  type: types.SET_ONBOARDING_COMPLETION,
  isCompleted,
});

export const startListenConnectionState = () => ({
  type: types.START_LISTEN_CONNECTION_STATE,
});

export const stopListenConnectionState = () => ({
  type: types.STOP_LISTEN_CONNECTION_STATE,
});

export const setConnectionState = (isConnected) => ({
  type: types.SET_CONNECTION_STATE,
  isConnected,
});
