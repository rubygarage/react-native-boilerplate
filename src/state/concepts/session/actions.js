import * as types from './types';

export const setTokens = (tokens) => ({
  type: types.SET_TOKENS,
  tokens,
});

export const removeTokens = () => ({
  type: types.SET_TOKENS,
  tokens: null,
});

export const setUserId = (id) => ({
  type: types.SET_USER_ID,
  id,
});
