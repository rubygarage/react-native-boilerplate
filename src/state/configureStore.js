import { createStore, applyMiddleware } from 'redux';
import { createLogicMiddleware } from 'redux-logic';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore, persistReducer } from 'redux-persist';
import { Settings } from 'luxon';
import AsyncStorage from '@react-native-community/async-storage';

import buildHttpClient from 'lib/httpClient';
import { rootHttpClient } from 'lib/httpClient/buildHttpClient';
import mountInterceptors from 'lib/httpClient/mountInterceptors';
import conceptsOperations from './concepts/operationsRoot';
import flashOperations from './flash/operations';
import rootReducer from './store';

const operations = [...conceptsOperations, ...flashOperations];

const configureStore = () => {
  const persistConfig = {
    key: 'root',
    storage: AsyncStorage,
    whitelist: ['data', 'application'],
    timeout: null,
  };
  const persistedReducer = persistReducer(persistConfig, rootReducer);

  const operationsDependencies = {
    httpClient: buildHttpClient(),
    rootHttpClient,
  };

  const logicMiddleware = createLogicMiddleware(operations, operationsDependencies);
  const middlewares = applyMiddleware(logicMiddleware);
  const enhancer = composeWithDevTools(middlewares);

  const store = createStore(persistedReducer, enhancer);
  const persistor = persistStore(store);
  store.logicMiddleware = logicMiddleware;
  store.httpClient = operationsDependencies.httpClient;

  mountInterceptors(store);

  if (module.hot) {
    module.hot.accept(() => {
      // eslint-disable-next-line global-require
      store.replaceReducer(require('./store').default);
    });
  }

  Settings.defaultLocale = store.getState().intl.locale;

  return { store, persistor };
};

export default configureStore;
