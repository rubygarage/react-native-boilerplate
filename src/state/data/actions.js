import * as types from './types';

export const dataApiRequest = ({ endpoint }) => ({
  type: types.DATA_API_REQUEST,
  endpoint,
});

export const dataApiSuccess = ({ endpoint, response }) => ({
  type: types.DATA_API_SUCCESS,
  endpoint,
  response,
});

export const dataApiFailure = ({ endpoint }) => ({
  type: types.DATA_API_FAILURE,
  endpoint,
});

export const dataDelete = ({ kind, ids }) => ({
  type: types.DATA_DELETE,
  kind,
  ids,
});
