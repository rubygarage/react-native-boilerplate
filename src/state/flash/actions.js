import { ALERT_TYPES } from 'constants/alert';
import * as types from './types';

export const showFlash = ({
  id,
  lifetime,
  priority,
  isSingleShot = false,
  flashProps,
}) => ({
  type: types.SHOW_FLASH,
  flash: {
    id,
    lifetime,
    priority,
    isSingleShot,
    flashProps,
  },
});

export const showErrorFlash = (action) => ({
  type: types.SHOW_FLASH,
  flash: {
    isSingleShot: false,
    flashProps: {
      title: { id: 'shared.oops' },
      type: ALERT_TYPES.error,
      action,
      actionTextId: 'shared.tryAgain',
      isCloseable: true,
    },
  },
});

export const addFlash = (flash) => ({
  type: types.ADD_FLASH,
  flash,
});

export const hideFlash = (id) => ({
  type: types.HIDE_FLASH,
  id,
});

export const removeFlash = (id, isSingleShot) => ({
  type: types.REMOVE_FLASH,
  id,
  isSingleShot,
});
