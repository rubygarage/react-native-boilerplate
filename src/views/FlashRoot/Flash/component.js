import React from 'react';
import PropTypes from 'prop-types';

import Alert from 'views/shared/Alert';

import useFlashMethods from './hooks/useFlashMethods';

const Flash = (props) => {
  const { flashProps } = props;

  const { onHideFlash, onActionClick } = useFlashMethods(props);

  return (
    <Alert
      {...flashProps}
      onCloseClick={flashProps.isCloseable ? onHideFlash : undefined}
      onActionClick={onActionClick}
    />
  );
};

Flash.propTypes = {
  doAnimation: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  flashProps: PropTypes.shape({
    action: PropTypes.shape({}),
    isCloseable: PropTypes.bool,
  }),
  lifetime: PropTypes.number,
};

Flash.defaultProps = {
  lifetime: undefined,
  flashProps: {},
};

export default Flash;
