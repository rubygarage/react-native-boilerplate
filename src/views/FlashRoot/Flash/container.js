import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { hideFlash as hideFlashAction } from 'state/flash/actions';
import Alert from 'views/shared/Alert';

class Flash extends Component {
  componentDidMount() {
    const { lifetime } = this.props;
    if (lifetime) {
      this.timerHandle = setInterval(this.onHideFlash, lifetime);
    }
  }

  componentWillUnmount() {
    clearInterval(this.timerHandle);
  }

  onHideFlash = () => {
    const { hideFlash, id, doAnimation } = this.props;
    hideFlash(id);
    doAnimation();
  }

  onActionClick = () => {
    const { dispatch, hideFlash, id, flashProps: { action } } = this.props;
    hideFlash(id);
    dispatch(action);
  }

  render() {
    const { flashProps } = this.props;
    const { isCloseable } = flashProps;
    return (
      <Alert
        {...flashProps}
        onCloseClick={isCloseable ? this.onHideFlash : undefined}
        onActionClick={this.onActionClick}
      />
    );
  }
}

Flash.propTypes = {
  hideFlash: PropTypes.func.isRequired,
  doAnimation: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  flashProps: PropTypes.shape({
    action: PropTypes.shape({}),
    isCloseable: PropTypes.bool,
  }).isRequired,
  lifetime: PropTypes.number,
};

Flash.defaultProps = {
  lifetime: undefined,
};

const mapDispatchToProps = (dispatch) => ({
  ...bindActionCreators({ hideFlash: hideFlashAction }, dispatch),
  dispatch,
});

export { Flash as FlashContainer };
export default connect(null, mapDispatchToProps)(Flash);
