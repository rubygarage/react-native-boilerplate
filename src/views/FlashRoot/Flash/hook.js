import { useRef, useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { hideFlash } from 'state/flash/actions';

const useContainer = (props) => {
  const timerHandle = useRef();

  const dispatch = useDispatch();

  const {
    id,
    flashProps: { action },
    doAnimation,
    lifetime,
  } = props;

  const onHideFlash = () => {
    dispatch(hideFlash(id));
    doAnimation();
  };

  const onActionClick = () => {
    dispatch(hideFlash(id));
    dispatch(action);
  };

  const didMount = () => {
    if (lifetime) {
      timerHandle.current = setInterval(onHideFlash, lifetime);
    }

    return () => {
      clearInterval(timerHandle.current);
    };
  };

  useEffect(didMount, []);

  return {
    onHideFlash,
    onActionClick,
    didMount,
  };
};

export default useContainer;
