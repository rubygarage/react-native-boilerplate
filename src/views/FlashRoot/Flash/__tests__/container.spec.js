import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { DateTime } from 'luxon';

import { hideFlash } from 'state/flash/actions';
import diveTo from 'utils/testHelpers/diveToEnzyme';
import Flash, { FlashContainer } from '../container';

jest.useFakeTimers();

jest.mock('state/flash/actions', () => ({
  hideFlash: jest.fn(),
}));

describe('Flash container', () => {
  const store = configureStore()({});
  store.dispatch = jest.fn();
  const id = '300';
  const doAnimation = jest.fn();
  const action = { type: 'SOME_ACTION' };
  const flashProps = { action };
  const defaultProps = { id, doAnimation, flashProps, store };
  const options = { disableLifecycleMethods: true };
  const wrapper = shallow(<Flash {...defaultProps} />, options);
  const container = diveTo(wrapper, FlashContainer);
  const instance = container.instance();

  jest.spyOn(DateTime, 'local');

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('renders correctly closeable', () => {
    container.setProps({ ...defaultProps, flashProps: { ...flashProps, isCloseable: true } });
    expect(container).toMatchSnapshot();
  });

  it('componentDidMount updates timer and sets interval', () => {
    const onHideFlashSpy = jest.spyOn(instance, 'onHideFlash');

    const lifetime = 30 * 1000;
    container.setProps({ ...defaultProps, lifetime });
    instance.componentDidMount();

    expect(setInterval).toHaveBeenCalledWith(
      onHideFlashSpy, lifetime,
    );
  });

  it('componentDidMount does nothing without lifetime', () => {
    container.setProps({ ...defaultProps, lifetime: undefined });
    instance.componentDidMount();

    expect(setInterval).not.toHaveBeenCalled();
  });

  it('componentWillUnmount clears timerHandle interval', () => {
    instance.timerHandle = 'timerHandle mock';
    instance.componentWillUnmount();

    expect(clearInterval).toHaveBeenCalledWith('timerHandle mock');
  });

  it('onHideFlash hides flash', () => {
    instance.onHideFlash();

    expect(hideFlash).toHaveBeenCalledWith(id);
  });

  it('onActionClick dispatches action and hides modal', () => {
    instance.onActionClick();

    expect(hideFlash).toHaveBeenCalledWith(id);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
