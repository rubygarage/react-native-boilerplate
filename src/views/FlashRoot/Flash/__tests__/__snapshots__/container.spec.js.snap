// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flash container renders correctly 1`] = `
<WithTheme(Alert)
  action={
    Object {
      "type": "SOME_ACTION",
    }
  }
  onActionClick={[Function]}
/>
`;

exports[`Flash container renders correctly closeable 1`] = `
<WithTheme(Alert)
  action={
    Object {
      "type": "SOME_ACTION",
    }
  }
  isCloseable={true}
  onActionClick={[Function]}
  onCloseClick={[Function]}
/>
`;
